File: old/src/main/java/pie/backtracking/Permutations.java
package pie.backtracking;

/** Created by gakshintala on 6/10/16. */
public class Permutations {
  public static void main(String[] args) {
    var str = "abc";
    permute(str, new StringBuilder(), new boolean[str.length()]);
  }

  private static void permute(String str, StringBuilder permutation, boolean[] used) {
    // Check the satisfying condition
    if (permutation.length() == str.length()) {
      System.out.println(permutation);
      return;
    }
    for (var i = 0; i < str.length(); i++) {
      // Check is Valid
      if (!used[i]) {
        permutation.append(str.charAt(i));
        used[i] = true;
        permute(str, permutation, used);

        // Backtrack
        used[i] = false;
        permutation.setLength(permutation.length() - 1);
      }
    }
  }
}

File: src/backtracking/Permutations.kt
package backtracking

fun permutations(str: String, res: String, used: BooleanArray) {
    var result = res
    val strLength = str.length
    if (res.length == strLength) {
        println(res)
        return
    }
    for (i in 0 until strLength) {
        if (!used[i]) {
            result += str[i]
            used[i] = true
            permutations(str, result, used)
            used[i] = false
            result = result.dropLast(1)
        }
    }
}

fun main() {
    permutations("abc", "", BooleanArray(3))
}